src/effects/RealtimeEffectList.cpp

```diff
@@ -87,7 +88,10 @@ RealtimeEffectList::AddState(const PluginID &id)
 {
    auto pState = std::make_shared<RealtimeEffectState>(id);
    if (id.empty() || pState->GetEffect() != nullptr) {
-      mStates.emplace_back(pState);
+      auto shallowCopy = mStates;
+      shallowCopy.emplace_back(pState);
+      // Lock for only a short time
+      (LockGuard{ mLock }, swap(shallowCopy, mStates));
       return pState;
    }
    else
```

Comment:
What happens though if a call to `::RemoveState` arrives and completes just after `auto shallowCopy = mStates;` ?
We would end up with the new state added at the back, but the state which we thought we removed, would still be there.

As long as calls to ::AddState and ::RemoveState will come from the same thread (which I assume is true?) this should not happen.

I see the point of doing this scheme: we do not want to run the risk of the system reallocating the whole vector, under a lock. But what about calling .reserve() instead?

Reply:
You are correct that calls to AddState and RemoveState must be serialized in the main thread, and you will see I also add explicit comments in the header file about this expected usage.

I want to keep the scope of the lock guard as brief as possible, so there is only a swap of a few pointers implementing the swap of vectors, and therefore the need to create a second vector first, and therefore also the previous commit that changes management from unique to shared pointers.

But reserve() might do a free store reallocation while under the lock guard.